Mathlib/RingTheory/MvPowerSeries/Evaluation.lean

```diff
@@ -0,0 +1,437 @@
+    coeToMvPowerSeries_denseRange
+    (coeToMvPowerSeries_uniformContinuous hφ ha)
+
+theorem eval₂Hom_apply (hφ : Continuous φ) (ha : EvalDomain a) (f : MvPowerSeries σ R) :
+    eval₂Hom hφ ha f =
+      coeToMvPowerSeries_denseInducing.extend (MvPolynomial.eval₂ φ a) f :=
+  rfl
+
+theorem coe_eval₂Hom (hφ : Continuous φ) (ha : EvalDomain a) :
+    ⇑(eval₂Hom hφ ha) = eval₂ φ a := by
+  ext f
+  let hf := fun (p : MvPolynomial σ R) ↦ p = f
+  simp only [eval₂Hom_apply, eval₂]
+  split_ifs with h
+  · obtain ⟨p, rfl⟩ := h
+    simpa [MvPolynomial.coe_eval₂Hom, coeToMvPowerSeries.ringHom_apply] using
+      IsDenseInducing.extend_eq coeToMvPowerSeries_denseInducing
+        (coeToMvPowerSeries_uniformContinuous hφ ha).continuous p
+  · rfl
+
+theorem uniformContinuous_eval₂ (hφ : Continuous φ) (ha : EvalDomain a) :
+    UniformContinuous (eval₂ φ a) := by
+  rw [← coe_eval₂Hom hφ ha]
+  exact uniformContinuous_uniformly_extend
+    coeToMvPowerSeries_uniformInducing
+    coeToMvPowerSeries_denseRange
+    (coeToMvPowerSeries_uniformContinuous hφ ha)
+
+theorem continuous_eval₂ (hφ : Continuous φ) (ha : EvalDomain a) :
+    Continuous (eval₂ φ a : MvPowerSeries σ R → S) :=
+  (uniformContinuous_eval₂ hφ ha).continuous
+
+theorem hasSum_eval₂ (hφ : Continuous φ) (ha : EvalDomain a)(f : MvPowerSeries σ R) :
+    HasSum
+    (fun (d : σ →₀ ℕ) ↦ φ (coeff R d f) * (d.prod fun s e => (a s) ^ e))
+    (MvPowerSeries.eval₂ φ a f) := by
+  rw [← coe_eval₂Hom hφ ha, eval₂Hom_apply hφ ha]
+  convert (hasSum_of_monomials_self f).map (eval₂Hom hφ ha) (?_) with d
+  · simp only [Function.comp_apply, coe_eval₂Hom, ← MvPolynomial.coe_monomial,
+    eval₂_coe, eval₂_monomial]
+  · rw [coe_eval₂Hom]; exact continuous_eval₂ hφ ha
+
+theorem eval₂_eq_sum (hφ : Continuous φ) (ha : EvalDomain a)(f : MvPowerSeries σ R) :
+    MvPowerSeries.eval₂ φ a f =
+      tsum (fun (d : σ →₀ ℕ) ↦ φ (coeff R d f) * (d.prod fun s e => (a s) ^ e)) :=
+  (hasSum_eval₂ hφ ha f).tsum_eq.symm
+
+theorem eval₂_unique (hφ : Continuous φ) (ha : EvalDomain a)
+    {ε : MvPowerSeries σ R →+* S} (hε : Continuous ε)
+    (h : ∀ p : MvPolynomial σ R, ε p = MvPolynomial.eval₂ φ a p) :
+    ε = eval₂ φ a := by
+  rw [← coe_eval₂Hom hφ ha]
+  exact (MvPolynomial.coeToMvPowerSeries_denseInducing.extend_unique h hε).symm
+
+theorem comp_eval₂ (hφ : Continuous φ) (ha : EvalDomain a)
+    {T : Type*} [UniformSpace T] [CompleteSpace T] [T2Space T]
+    [CommRing T] [IsTopologicalRing T] [IsLinearTopology T T] [UniformAddGroup T]
+    {ε : S →+* T} (hε : Continuous ε) :
+    ε ∘ eval₂ φ a = eval₂ (ε.comp φ) (ε ∘ a) := by
+  rw [← coe_eval₂Hom hφ ha, ← coe_comp]
+  apply eval₂_unique _ (ha.comp hε)
+  · simp only [coe_comp, coe_eval₂Hom hφ ha]
+    exact Continuous.comp hε (continuous_eval₂ hφ ha)
+  · intro p
+    simp only [coe_comp, Function.comp_apply, eval₂_coe]
+    rw [coe_eval₂Hom hφ ha, eval₂_coe φ a, ← MvPolynomial.coe_eval₂Hom, ← comp_apply,
+    MvPolynomial.comp_eval₂Hom]
```

Comment:
```suggestion
  apply eval₂_unique _ (ha.comp hε)
  · exact Continuous.comp hε (continuous_eval₂ hφ ha)
  · intro p
    simp only [Function.comp_apply, eval₂_coe]
    rw [← MvPolynomial.coe_eval₂Hom, ← comp_apply, MvPolynomial.comp_eval₂Hom]
```

(This is allowed by the previous suggestion: you don't have to make it a morphism anymore)